runtime/vm/resolvesupport.cpp

```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 
 }
 
+j9object_t
+resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags)
+{
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramCP + cpIndex;
+	j9object_t value = ramCPEntry->value;
+
+	J9Class *ramClass = ramCP->ramClass;
+	J9ROMClass *romClass = ramClass->romClass;
+	J9ROMConstantDynamicRef *romConstantRef = (J9ROMConstantDynamicRef*)(J9_ROM_CP_FROM_CP(ramCP) + cpIndex);
+	J9SRP *callSiteData = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass);
+	U_16 *bsmIndices = (U_16 *) (callSiteData + romClass->callSiteCount);
+	U_16 *bsmData = bsmIndices + romClass->callSiteCount;
+	U_32 bsmIndex = romConstantRef->bsmIndexAndCpType >> J9DescriptionCpTypeShift;
+	J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+
+	J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+	J9Class *throwable = internalFindKnownClass(vmThread, J9VMCONSTANTPOOL_JAVALANGTHROWABLE, J9_FINDKNOWNCLASS_FLAG_INITIALIZE);
+	UDATA resolved = 0;
+
+retry:
+	/* Check if already resolved */
+	if (NULL != value) {
+		resolved = 1;
+	} else if (NULL != ramCPEntry->exception) {
+		if (isSameOrSuperClassOf(throwable, J9VM_J9CLASS_FROM_HEAPCLASS(vmThread, ramCPEntry->exception))) {
+			/* Resolved with exception */
+			VM_VMHelpers::setExceptionPending(vmThread, ramCPEntry->exception);
+			resolved = 1;
+		} else if (ramCPEntry->exception == vmThread->javaVM->voidReflectClass->classObject) {
+			/* Resolved with null reference */
+			resolved = 1;
+		} else if (ramCPEntry->exception != vmThread->threadObject) {
+			/* Another thread is currently trying to resolve the entry */
+			// yield(vmThread);
```

Comment:
VM_AtomicSupport::yieldCPU();

Reply:
Actually, yield my not be sufficient to ensure that the resolving thread gets precedence.

You can add a monitor (ie. condyMonitor) to each j9vmthread. Then the thread can wait on `((J9VMThread *)J9VMJAVALANGTHREAD_THREADREF(ramCPEntry->exception))->condyMonitor`.

And the code that writes value can notify condyMonitor